### ISSUE TITLE
Blocking: Tested working install instructions.

### ISSUE THREAD

**Issue by taqtiqa-mark** [gus0k/pymarket#15]
* pymarket version: latest
* Python version: 3.6
* Operating System: Ubuntu Bionic 

### Description

I followed the install instructions.

However, after upgrading pip via `pip install --user --upgrade pip` pip refuses to start and raises this instead:

```bash
Traceback (most recent call last):
  File "/usr/bin/pip3", line 9, in <module>
    from pip import main
ImportError: cannot import name 'main'
```

This apparently a well known thing, previously reported.
- https://stackoverflow.com/questions/49836676/error-after-upgrading-pip-cannot-import-name-main
- https://stackoverflow.com/questions/28210269/importerror-cannot-import-name-main-when-running-pip-version-command-in-windo
- https://askubuntu.com/questions/1108070/pip3-command-fails-after-upgrading-to-pip-version-18-1
- https://askubuntu.com/questions/1135252/pip-broken-after-upgrading-to-19-04
- https://askubuntu.com/questions/1025189/pip-is-not-working-importerror-no-module-named-pip-internal
- https://askubuntu.com/questions/1025793/running-pip3-importerror-cannot-import-name-main

**I hope you can appreciate how frustrating this is?  You have broken a users pip installation on their computer!!**

Given that you appear to refuse to test your instructions and just wave your hand that this works everywhere:

The only way I can close this issue is:
1. Update the vagrant scripts such that they work.  then update your install instructions accordingly.
2. Remove all references to install instructions for Linux, and state the only way to get this to run is to use the docker container you provide, and give the link to the built container.

As I commented [here](https://github.com/openjournals/joss-reviews/issues/1591#issuecomment-545712658), the fastest way to publish this package is the second of the above options.

### What I Did

```bash
git clone https://github.com/gus0k/pymarket
pushd pymarket/dev/bionic
vagrant up
```

**Comment by gus0k** [comment]
@taqtiqa-mark I'm sorry about the error in the installation instructions. I haven't been able to set up the digital ocean account (which as far as i understand are the images your are using), so I could not reproduce your results exactly. 

I understood what was pushing the library to python version 3.6> and I'm gonna try to make some changes in order to make it as easy to install as possible. If this fails, we can resort to using your Docker approach.

I should have something for you in a couple of days. I'll update you as soon as possible.

**Comment by gus0k** [comment]
@taqtiqa-mark 

I made some changes that might could solve your installation issues. The build is working on travis but my internet connection is pretty bad and I won't be able download new dockers to test it there. 

If you could try the following, it can help me understand if this changes are in the right direction:

Without updating pip or anything, try to run the following (change python to python3 if you have both available), or build from the branch "python 3.5":

```sh
python -m pip install https://github.com/gus0k/pymarket/archive/python3.5.zip --user
```

Basically, I made the library compatible with python 3.5 and decreased the minimum requirements it has. I think this should suffice to solve the installation issues as it runs on the default configuration of most Ubuntus.

Looking forward to hear from you,
Diego

**Comment by taqtiqa-mark** [comment]
@gus0k Thanks for making these changes.  A Digital Ocean account is not necessary: VirtualBox or VM Ware will allow you to obtain a fresh Linux installation (whichever you think you can get working) to start working from.

Please either install some VM package and **_you_** develop and test **_your_** install instructions.

Alternatively make the docker image that you claim works available, and remove references to linux installations and instructions.

**Comment by gus0k** [comment]
@taqtiqa-mark I tested it on a local vagrant instalation and it's working.

In branch python3.5, inside dev/diego you can find to vagrantfiles, one for bionic and one for xenial (one is running python3.5 and the other python 3.6).

The code I ran is the following

```sh
vagrant up
vagrant provision
vagrant reload
vagrant ssh
python3 -m pip install https://github.com/gus0k/pymarket/archive/python3.5.zip --user
```
The last line I ran it inside the vagrant because for some reason I wasn't able to install anything (not only pymarket) from the provision line.

It works for both systems. Right now works for python >= 3.5, no need to update pip.

**Comment by taqtiqa-mark** [comment]
Any reason you don't run the test suite?
Can you pipe the output showing clean runs.

Right now this will only expected/shown to work on Ubuntu Bionic 18.04 correct?

**Comment by gus0k** [comment]
@taqtiqa-mark , it's done now.

You can take a look at both provision.log files in the corresponding folders.

Right now it has been tested in: my local pc, in all the travis you can check from the main page and two vagrant files:

# dev/diego/bionic

Ubuntu 18.04.03, Python 3.6.8, 9.0.1

# dev/diego/xenial

Ubuntu 16.04.0.1, Python 3.5.2, pip 8.1.1

**Comment by taqtiqa-mark** [comment]
Can you let me kown when you have merged this into master and updated your instructions to reflect what is in those scripts - in a brief read I noted some package installs that are new.

**Comment by gus0k** [comment]
Done

**Comment by taqtiqa-mark** [comment]
Hi,  I consulted:
https://pymarket.readthedocs.io/en/master/installation.html

Please do the following:
1. Start a some fresh OS, which ever you say your instructions work for.
2. Read the instructions on the above page
3. As you encounter an instruction,
   a. type it into the console
   b. press enter.

I believe you will discover your instructions are incomplete.

**Comment by gus0k** [comment]
I just tried in my friend's PC and it works. Just in case, you should not write the parenthesis..

-------- Original Message --------

**Comment by taqtiqa-mark** [comment]
>  I just tried in my friend's PC and it works. 

How can that be true?:

![image](https://user-images.githubusercontent.com/1468258/68550779-34586f00-045a-11ea-986f-cc72f9be116b.png)

**Comment by gus0k** [comment]
That seems to be an old version. You need to access the docs built against master, the other only updates with a version bump, which I haven't done.

https://pymarket.readthedocs.io/en/latest/installation.html

-------- Original Message --------

**Comment by taqtiqa-mark** [comment]
Sorry I've wasted enough time on this. Let me know when you've updated the submission.

**Comment by gus0k** [comment]
Done

**Comment by taqtiqa-mark** [comment]
Apologies for the terse reply above.

I think I understand what you've done: You're Vagrant files are the setup for installing from source.  
Correct?

If so, your instructions for the "From Sources" (https://pymarket.readthedocs.io/en/latest/installation.html#from-sources) section should be :

For Ubuntu 16.04 & 18.04:
```bash
apt-get install --yes  python3-pip=9.0.1-2.3~ubuntu1.18.04.1
apt-get install --yes pkg-config
apt-get install --yes libfreetype6-dev
apt-get install --yes libpng12-dev
python3 -m pip -V
echo "Installing setupttools"
python3 -m pip install 'setuptools>=27.3'
echo "Checking setuptools"
python3 -m pip freeze | grep setuptools
 
cd /~/src
git clone -b python3.5 https://github.com/gus0k/pymarket.git
cd pymarket
python3 setup.py install
python3 setup.py test
```

The broken pip installation that I report in this ticket is from following the stable release instructions.
Can you please workout stable release install instructions as detailed as the above.
Once you have worked that out, please add them to the existing Vagrantfile(s) I provided.

**Comment by gus0k** [comment]
Hello,

Sorry for my late replay. I have been busy with a deadline. 
To make it easy to maintain, I erased all documentations but latest version, there is no longer a stable one. In the future, if we have many versions we might consider brining it back up.

Without wanting to sound rude, I have provided two working vagrant files as you requested at some point. Even If I modify your vagrant files, because the base image is different, the needs might be different. Because I cannot run it, I would probably just break it more. 

Instructions as they are know work fine for anyone that wants to use the library, and they are a guide (testes in Ubuntu 16/18) on how to run also the tests. 

Feel free to try those instructions in your own images. If that does not work, we can try to figure out why it does not for each particualar case.

Best regards, 
DIego

**Comment by taqtiqa-mark** [comment]
I think you've misunderstood the documentation instructions that this issue relates to.

I am referring to the instructions as at this date, at this URL:
https://pymarket.readthedocs.io/en/latest/installation.html

The Vagrant files I provided tried to get to the point where the following command would run (these instruction is still there as at the time of writing):

`python -m pip install pymarket --user`
 
In response you provided Vagrant files that get a user to the point that the following command would run:

`python3 setup.py install`

**Comment by gus0k** [comment]
Have you tried your Vagrant files again with the new version of pymarket? If not, please try that:
I think that they should be working now as they are.
If they do not, please let me know because there might be an issue I have not seen yet.

Best regards,
Diego

-------- Original Message --------

**Comment by taqtiqa-mark** [comment]
If you merge the pull request and run:
```
pushd dev/xenial
vagrant up --provider=virtualbox
```
 and 

```
pushd dev/xenial
vagrant up --provider=virtualbox
```

I believe you will see that both are broken in different ways.  

I suggest getting just one (bionic?) to work, and remove references to Xenial, Mandriva, etc. until you have added a build I can verify works - sorry I can't any longer accept assurances that things work given the history of those claims.

**Comment by gus0k** [comment]
As far as I can see from your install.sh files your are running the tests as
`make test`

when the updated version of the installation uses (https://pymarket.readthedocs.io/en/latest/installation.html):
`pytest`

Also, the requeriments_dev.txt is not refered in the documentation and is clearly outdated. I'll remove it. If you want to run the test the way you are doing it you most likely only need to install:

`python3 -m pip install pytest --user`

Regarding Vagrant, for some reason, my computer is not running any vagrant at all. I'm running into several errors which do not seem to be easily fixed.
If you insist on having the tests run in Vagrant (as opposed to Docker), It would help me to have the logs that you obtain.

Best regards,
Diego

**Comment by taqtiqa-mark** [comment]
> Also, the requeriments_dev.txt is not refered in the documentation and is clearly outdated. 

Yes please remove it if you no longer recommend its use.  It is still referenced in your original docker file, here: 
https://github.com/gus0k/pymarket/blob/master/dev/Dockerfile#L12
which was the starting point to verify the instructions.

I haven't been able to reproduce your steps using a VM. Aside from installing a fresh OS on my laptop, a VM is the only way to verify your instructions work for some other user.  
And the only way to ensure your instructions don't break my work machine - I showed your earlier instructions would have broken my pip installation - happily I dodged that bullet.

Vagrant is the way I've launched a VM, as a courtesy I contributed those scripts.  Unfortunately that didn't work out. 
There was, and is, nothing stopping you from testing your instructions on something such as VirtualBox or some such environment.

**Due to non-responses. I'm abandoning the effort to verify your installation instructions in a classical user environment.**

I will now try to verify the docker images build and run.

Please remove any reference to pymarket being able to be installed and run on anything other than docker images.

**Comment by gus0k** [comment]
I just downloaded Ubuntu 16 and 18 from the official website and created two virtual machines.

After running normal updates: `sudo apt-get update` and installing pip `sudo apt-get install python3-pip`,

I was able to run the installation instructions without any problem:

```sh
python3 -m pip install pymarket
```
to test it:

```sh
git clone https://github.com/gus0k/pymarket
cd pymarket
python3 -m pytest
```

And it worked like a charm.

**Comment by taqtiqa-mark** [comment]
Please see the updated pull request.  Please make changes to this scripts that reproduce the following failure:

> ... it worked like a charm.


```
<snip>
pymarket1604: Successfully installed cycler-0.10.0 decorator-4.4.1 kiwisolver-1.1.0 matplotl
ib-2.2.4 networkx-2.2 numpy-1.17.4 pandas-0.23.0 pulp-1.6.10 pymarket-0.7.6 pyparsing-2.4.5 pyth
on-dateutil-2.8.1 pytz-2019.3
    pymarket1604: You are using pip version 8.1.1, however version 19.3.1 is available.
    pymarket1604: You should consider upgrading via the 'pip install --upgrade pip' command.
    pymarket1604: Cloning into 'pymarket'...
    pymarket1604: /usr/bin/python3: No module named pytest
```

**Comment by gus0k** [comment]
It seems that you haven't installed pytest and it seems that I haven't added it in the documentation. I'm truly sorry about that. In the comment above, I meant pytest instead of pymarket in the first line of the testing procedure (typing error)

```sh
python3 -m pip install pytest
```
If you install pymarket from source, then it gets installed automatically (that is what is described in the documentation). If you mix both installations (from pip and then clone), you need to install pytest additionaly (that case is not covered in the documentation yet, but works if you install pytest on its own).

**Comment by taqtiqa-mark** [comment]
No worries.
Please make the changes you say are needed when you merge the current pull request.

I hope you understand why I won't be able to a close these issues while OS's are cited as working, but for which there is no evidence.  Evidence would be:
 - scripts/instructions I have tested 
-  the other reviewer states they have tested OS xyz.

**Comment by gus0k** [comment]
Hello,

I merged your pull request and added the line missing to install pytest. I also updated the documents to reflect the fact that pytest needs to be installed to run tests.

I think that your test should run now, unless something strange occurs.

Thank you for your time,
Diego

**Comment by gus0k** [comment]
@taqtiqa-mark Have you had the opportunity to test your vagrant files with my modifications?
